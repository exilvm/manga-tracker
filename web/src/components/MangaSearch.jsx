import React, { useCallback, useState } from 'react';
import {
  IconButton,
  InputBase,
  Popper,
} from '@material-ui/core';
import { fade, makeStyles } from '@material-ui/core/styles';
import Autocomplete from '@material-ui/lab/Autocomplete';
import PropTypes from 'prop-types';

import { Search as SearchIcon } from '@material-ui/icons';

import NextLink from 'next/link';
import throttle from 'lodash.throttle';

const useStyles = makeStyles((theme) => ({
  search: {
    position: 'relative',
    borderRadius: theme.shape.borderRadius,
    backgroundColor: fade(theme.palette.common.white, 0.15),
    '&:hover': {
      backgroundColor: fade(theme.palette.common.white, 0.25),
    },
    width: 'auto',
    [theme.breakpoints.down(400)]: {

    },
  },
  inputRoot: {
    color: 'inherit',
  },
  inputInput: {
    width: '100%',
    marginLeft: '1em',
    transition: theme.transitions.create('width'),
    [theme.breakpoints.up('sm')]: {
      width: '16em',
      '&:focus': {
        width: '24em',
      },
    },
  },
  popper: {
    zIndex: theme.zIndex.modal,
    marginTop: '10px',
    width: '200px !important',
    [theme.breakpoints.up('sm')]: {
      width: '450px !important',
    },
  },
}));


export default function MangaSearch(props) {
  const {
    placeholder = 'Searchâ€¦',
    renderItem,
    inputClasses = {},
    popperProps = {},
    id = 'manga-search',
  } = props;

  const [value, setValue] = useState('');
  const [options, setOptions] = useState([]);
  const classes = useStyles();

  const handleChange = useCallback((event) => {
    setValue(event.target.value);
  }, []);

  const throttleFetch = React.useMemo(
    () => throttle((query, cb) => {
      fetch('/api/quicksearch?query=' + encodeURIComponent(query))
        .then(res => res.json())
        .then(js => cb(js))
        .catch(err => {
          console.error(err);
          cb(null);
        });
    }, 200),
    []
  );

  React.useEffect(() => {
    let active = true;
    if (value.length < 2) {
      setOptions([]);
      return undefined;
    }

    throttleFetch(value, (results) => {
      if (active) {
        setOptions(results || []);
      }
    });

    return () => {
      active = false;
    };
  }, [value, throttleFetch]);

  const renderListOption = renderItem ||
    ((option) => (
      <NextLink href='/manga/[id]' as={`/manga/${option.manga_id}`} prefetch={false}>
        <div>{option.title}</div>
      </NextLink>
    ));

  const BottomEndPopper = (pProps) => (
    <Popper {...pProps} placement='bottom-end' {...popperProps}>
      {pProps.children}
    </Popper>
  );

  return (
    <Autocomplete
      options={options}
      renderOption={renderListOption}
      clearOnBlur={false}
      getOptionLabel={(option => option.title)}
      filterOptions={(o) => o}
      freeSolo
      openOnFocus
      fullWidth
      id={id}
      PopperComponent={BottomEndPopper}
      classes={{
        popper: classes.popper,
        root: classes.search,
      }}
      renderInput={(params) => (
        <InputBase
          aria-label='Manga search'
          inputProps={params.inputProps}
          ref={params.InputProps.ref}
          onChange={handleChange}
          placeholder={placeholder}
          classes={{
            input: classes.inputInput,
            root: classes.inputRoot,
            ...inputClasses,
          }}
          endAdornment={(
            <IconButton>
              <SearchIcon />
            </IconButton>
          )}
        />
      )}
    />
  );
}

MangaSearch.propTypes = {
  placeholder: PropTypes.string,
  renderItem: PropTypes.func,
  inputClasses: PropTypes.object,
  popperProps: PropTypes.object,
};
