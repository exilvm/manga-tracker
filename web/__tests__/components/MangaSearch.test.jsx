import React from 'react';
import { act } from 'react-dom/test-utils';
import { createMount } from '@material-ui/core/test-utils';
import fetchMock from 'fetch-mock';
import { createSerializer } from 'enzyme-to-json';

import MangaSearch from '../../src/components/MangaSearch';

fetchMock.config.overwriteRoutes = true;
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));


describe('Search should render correctly', () => {
  it('without input', () => {
    const tree = createMount()(<MangaSearch />);

    expect(tree).toMatchSnapshot();
  });

  it('with valid input', async () => {
    const mockResult = [
      {
        manga_id: 1,
        title: 'Test 1',
      },
      {
        manga_id: 2,
        title: 'Test 2',
      },
      {
        manga_id: 3,
        title: 'Test 3',
      },
    ];
    const searchFn = jest.fn().mockImplementation(() => mockResult);

    fetchMock.mock('glob:/api/quicksearch?query=*', searchFn);

    const wrapper = createMount()(
      <MangaSearch id='test-input-id' />
    );

    // Find search input
    const input = wrapper.find('input#test-input-id');
    expect(input).toHaveLength(1);

    // Simulate text changes and test that the quicksearch endpoint wasn't called
    await act(async () => {
      input.simulate('change', { target: { value: 'test search' }});
    });
    expect(searchFn).toHaveBeenCalledTimes(1);
    wrapper.update();

    const listItems = wrapper.find('li');
    expect(listItems).toHaveLength(mockResult.length);
    // Order is important as the first result is the most likely
    expect(listItems.map(l => l.key()))
      .toEqual(mockResult.map(r => r.manga_id.toString()));

    expect(wrapper).toMatchSnapshot();
  });
});

describe('Search should behave correctly with user input', () => {
  it('Should not do requests of under 3 characters', () => {
    const searchFn = jest.fn().mockImplementation(() => []);

    fetchMock.mock('glob:/api/quicksearch?query=*', searchFn);

    const wrapper = createMount()(
      <MangaSearch id='test-input-id' />
    );

    // Find search input
    const input = wrapper.find('input#test-input-id');
    expect(input).toHaveLength(1);

    // Simulate text changes and test that the quicksearch endpoint wasn't called
    input.simulate('change', { target: { value: 'x' }});
    input.simulate('change', { target: { value: 'ab' }});
    input.simulate('change', { target: { value: 'a' }});
    input.simulate('change', { target: { value: '' }});

    expect(searchFn).toHaveBeenCalledTimes(0);
  });

  it('Should do a request with 3 or more characters', async () => {
    const searchFn = jest.fn().mockImplementation(() => []);

    fetchMock.mock('glob:/api/quicksearch?query=*', searchFn);

    const wrapper = createMount()(
      <MangaSearch id='test-input-id' />
    );

    // Find search input
    const input = wrapper.find('input#test-input-id');
    expect(input).toHaveLength(1);

    // Simulate text changes and test that the quicksearch endpoint wasn't called
    await act(async () => {
        input.simulate('change', { target: { value: 'abc' }});
    });
    expect(searchFn).toHaveBeenCalledTimes(1);
  });

  it('Should throttle fast requests', async () => {
    const searchFn = jest.fn().mockImplementation(() => []);

    fetchMock.mock('glob:/api/quicksearch?query=*', searchFn);

    const wrapper = createMount()(
      <MangaSearch id='test-input-id' />
    );

    // Find search input
    const input = wrapper.find('input#test-input-id');
    expect(input).toHaveLength(1);

    // Simulate text changes and test that the quicksearch endpoint wasn't called
    await act(async () => {
      input.simulate('change', { target: { value: 'abc' }});
      input.simulate('change', { target: { value: 'abcd' }});
      input.simulate('change', { target: { value: 'abcde' }});
    });
    expect(searchFn).toHaveBeenCalledTimes(1);
  });
});
