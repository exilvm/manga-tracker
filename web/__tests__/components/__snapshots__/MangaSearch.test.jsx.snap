// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly without input 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root makeStyles-search-1 MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <div
    aria-label="Manga search"
    className="MuiInputBase-root makeStyles-inputRoot-2 MuiInputBase-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-activedescendant={null}
      aria-autocomplete="list"
      aria-controls={null}
      autoCapitalize="none"
      autoComplete="off"
      className="MuiInputBase-input makeStyles-inputInput-3 MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
      disabled={false}
      id="test-id"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      placeholder="Searchâ€¦"
      spellCheck="false"
      type="text"
      value=""
    />
    <button
      className="MuiButtonBase-root MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
          />
        </svg>
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
